streaming/base/stream.py

```diff
@@ -0,0 +1,312 @@
+            are cached while they are being used. Uses a temp directory if not set. Defaults to
+            ``None``.
+        split (str, optional): Which dataset split to use, if any. If provided, we stream from/to
+            the ``split`` subdirs of  ``remote`` and ``local``. Defaults to ``None``.
+        proportion (float, optional): How much to upsample or downsample this dataset, as the
+            proportion of the total combined dataset that consists of this sub-dataset. We resample
+            each dataset accordingly to achieve those proportions. If provided, must be positive.
+            If set by any of them, must be set by all of them. Defaults to ``None``.
+        repeat (float, optional): How much to upsample or downsample this sub-dataset, as a
+            multipler on the number of samples. If provided, must be positive. Defaults to
+            ``None``.
+        samples (int, optional): How much to upsample or downsample this sub-dataset, as the exact
+            number of samples. If provided, must be positive. Defaults to ``None``.
+        download_retry (int, optional): Number of download re-attempts before giving up. Defaults
+            to ``None``.
+        download_timeout (float, optional): Number of seconds to wait for a shard to download
+            before raising an exception. Defaults to ``None``.
+        validate_hash (str, optional): Optional hash or checksum algorithm to use to validate
+            shards. Defaults to ``None``.
+        keep_zip (bool, optional): Whether to keep or delete the compressed form when decompressing
+            downloaded shards. If ``False``, keep iff remote is local or no remote. Defaults to
+            `None``.
+        keep_raw (bool, optional): Whether to keep or delete the decompressed form (or only form)
+            of shards after all their samples have been yielded this epoch. If ``False``, keep iff
+            remote is local or no remote and no compression. Defaults to ``None``.
+    """
+
+    def __init__(self,
+                 *,
+                 remote: Optional[str] = None,
+                 local: Optional[str] = None,
+                 split: Optional[str] = None,
+                 proportion: Optional[float] = None,
+                 repeat: Optional[float] = None,
+                 samples: Optional[int] = None,
+                 download_retry: Optional[int] = None,
+                 download_timeout: Optional[float] = None,
+                 validate_hash: Optional[str] = None,
+                 keep_zip: Optional[bool] = None,
+                 keep_raw: Optional[bool] = None) -> None:
+        if not (remote or local):
+            raise ValueError('Remote and/or local path must be provided')
+
+        self.remote = remote
+
+        self._local = local
+        self.local = local or mkdtemp()
+
+        self.split = split or ''
+
+        if proportion is not None and repeat is not None:
```

Comment:
the three options are all incompatible with each other, right? not just these two?

Reply:
fixed